File: app/apps/terminal/terminal_stream_protocol.mjs
import { Buffer } from 'node:buffer';
import { requireTerminalNodeModule } from './terminal_node_modules.mjs';

const { decode, encode } = requireTerminalNodeModule('@msgpack/msgpack');

export const TERMINAL_STREAM_CODEC = 'msgpack-v1';
export const MAX_TERMINAL_FRAME_BYTES = 32 * 1024 * 1024;

export function isRecord(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

export function asBytes(value) {
  if (value instanceof Uint8Array) {
    return value;
  }
  if (value instanceof ArrayBuffer) {
    return new Uint8Array(value);
  }
  return null;
}

export function encodePipeFrame(message) {
  const payload = Buffer.from(encode(message, { ignoreUndefined: true }));
  if (payload.byteLength > MAX_TERMINAL_FRAME_BYTES) {
    throw new RangeError(`terminal frame exceeds ${MAX_TERMINAL_FRAME_BYTES} bytes`);
  }
  const header = Buffer.allocUnsafe(4);
  header.writeUInt32BE(payload.byteLength, 0);
  return Buffer.concat([header, payload]);
}

export class PipeFrameDecoder {
  constructor(maxFrameBytes = MAX_TERMINAL_FRAME_BYTES) {
    this.maxFrameBytes = maxFrameBytes;
    this.buffer = Buffer.alloc(0);
  }

  push(chunk) {
    const next = Buffer.from(chunk);
    if (next.byteLength === 0) {
      return [];
    }
    this.buffer = this.buffer.byteLength === 0
      ? next
      : Buffer.concat([this.buffer, next]);

    const messages = [];
    while (this.buffer.byteLength >= 4) {
      const payloadLength = this.buffer.readUInt32BE(0);
      if (payloadLength === 0 || payloadLength > this.maxFrameBytes) {
        throw new RangeError(`invalid terminal frame length: ${payloadLength}`);
      }
      const frameLength = 4 + payloadLength;
      if (this.buffer.byteLength < frameLength) {
        break;
      }
      const payload = this.buffer.subarray(4, frameLength);
      const decoded = decode(payload);
      if (!isRecord(decoded)) {
        throw new TypeError('terminal frame must decode to an object');
      }
      messages.push(decoded);
      this.buffer = this.buffer.subarray(frameLength);
    }
    return messages;
  }
}

File: app/apps/terminal/terminal_stream_protocol.py
from collections.abc import Mapping
import struct
from typing import cast, final

import msgspec


TERMINAL_STREAM_CODEC = "msgpack-v1"
MAX_TERMINAL_FRAME_BYTES = 32 * 1024 * 1024
PIPE_FRAME_HEADER = struct.Struct(">I")

JsonObject = dict[str, object]
_ENCODER = msgspec.msgpack.Encoder()
_DECODER = msgspec.msgpack.Decoder(object)


def _as_object(raw: object) -> JsonObject | None:
    if not isinstance(raw, Mapping):
        return None
    payload: JsonObject = {}
    for key, value in cast(Mapping[object, object], raw).items():
        if isinstance(key, str):
            payload[key] = value
    return payload


def pack_message(payload: Mapping[str, object]) -> bytes:
    encoded = _ENCODER.encode(dict(payload))
    if len(encoded) > MAX_TERMINAL_FRAME_BYTES:
        raise ValueError("Terminal message exceeds the configured frame limit")
    return encoded


def unpack_message(payload: bytes) -> JsonObject:
    if not payload or len(payload) > MAX_TERMINAL_FRAME_BYTES:
        raise ValueError("Invalid terminal MessagePack payload size")
    try:
        decoded = _DECODER.decode(payload)
    except msgspec.DecodeError as exc:
        raise ValueError("Invalid terminal MessagePack payload") from exc
    message = _as_object(decoded)
    if message is None:
        raise ValueError("Terminal MessagePack payload must be an object")
    return message


def encode_pipe_message(payload: Mapping[str, object]) -> bytes:
    encoded = pack_message(payload)
    return PIPE_FRAME_HEADER.pack(len(encoded)) + encoded


@final
class PipeFrameDecoder:
    def __init__(self, max_frame_bytes: int = MAX_TERMINAL_FRAME_BYTES) -> None:
        self._buffer: bytearray = bytearray()
        self._max_frame_bytes: int = max_frame_bytes

    def push(self, chunk: bytes | bytearray) -> list[JsonObject]:
        self._buffer.extend(chunk)
        messages: list[JsonObject] = []
        while len(self._buffer) >= PIPE_FRAME_HEADER.size:
            (payload_length,) = cast(
                tuple[int],
                PIPE_FRAME_HEADER.unpack_from(self._buffer),
            )
            if payload_length <= 0 or payload_length > self._max_frame_bytes:
                raise ValueError(f"Invalid terminal pipe frame length: {payload_length}")
            frame_length = PIPE_FRAME_HEADER.size + payload_length
            if len(self._buffer) < frame_length:
                break
            payload = bytes(self._buffer[PIPE_FRAME_HEADER.size:frame_length])
            del self._buffer[:frame_length]
            messages.append(unpack_message(payload))
        return messages
